### ISSUE TITLE
kafka-node intigration with IBM Event Stream (message hub) failed

### ISSUE THREAD

**Issue by ddhabak1** [SOHU-Co/kafka-node#1188]
## Questions?

Trying to connect kafka-node with IBM Cloud Event Stream (message hub) using Node.js

var client = new Client({
	kafkaHost: 'kafka03-xxxx:1234, kafka05-xxxx:1234,kafka01--xxxx:1234,kafka04--xxxx:1234,kafka02--xxxx:1234',
	sasl: { mechanism: 'plain', username: 'foo', password: 'bar' }
});

var producer = new Producer(client, { requireAcks: 1 });

producer.on('ready', function () {
	var message = 'a message';
	var keyedMessage = new KeyedMessage('keyed', 'a keyed message');

	producer.send([{ topic: 'MY_TOPIC', partition: 1, messages: [message, keyedMessage], attributes: 1 }], function (
	  err,
	  result
	) {
	  console.log(err || result);
	  process.exit();
	});
  });

  producer.on('error', function (err) {
	console.log('error', err);
  });

var topics = [{ topic: 'MT_TOPIC', partition: 1 }];
var options = { autoCommit: true, fetchMaxWaitMs: 1000, fetchMaxBytes: 1024 * 1024 };

var consumer = new Consumer(client, topics, options);

consumer.on('message', function (message) {
	console.log(message);
});

consumer.on('error', function (err) {
	console.log('error', err);
});

I am getting request timeout . Whats wrong with the code.
Thanks for your contribution!

**Comment by hyperlink** [comment]
Is the connection SSL enabled? If so you need to let the KafkaClient know.

**Comment by ddhabak1** [comment]
Issue resolved.

**Comment by goemondo** [comment]
How did you solve it @ddhabak1 ? I got a same error even when SSL connection was enabled, which might be. I added `sslOptions` with an object `{ rejectUnauthorized: false }` and `ssl: true` in the option to `KafkaClient`.

**Comment by goemondo** [comment]
solved it by adding `servername: ${one_of_kafkaHost}` into `sslOptions`.

**Comment by ericqqqqq** [comment]
Hey, @goemondo  Do you mind posting your sample code? I am having the same issue.

```
const kafkaConfig = {
    kafkaHost: kafka_brokers_sasl,
    ssl: true,
    sslOptions: { rejectUnauthorized: false, servername: kafka_brokers_sasl },
    sasl: {
      mechanism: "plain",
      username: user,
      password: password,
    },
  };

const client = new kafka.KafkaClient(kafkaConfig)
```
I guess the error may be related to the recent updates from IBM Event Streams, therefore, I suppose that the same problem is causing this issue.

Thanks.

**Comment by goemondo** [comment]
@ericqqqqq Mine was just workaround so that not sure this is right answer.

```javascript
// credentials here is object from IBM Event Streams service credentials
const client = new kafka.KafkaClient({
  kafkaHost: credentials.kafka_brokers_sasl.join(','),
  sslOptions: {
    rejectUnauthorized: false,
    servername: credentials.kafka_brokers_sasl[0].match(/(.*):.*/)[1],
  },
  sasl: {
    mechanism: 'plain',
    username: credentials.user,
    password: credentials.api_key,
  }
});
```

`kafka-node` just passes this option to tls.connect() [here](https://github.com/SOHU-Co/kafka-node/blob/26fac1c16353b906e610def5fba4380778574339/lib/kafkaClient.js#L717), and the [node document](https://nodejs.org/api/tls.html#tls_tls_connect_options_callback) says

> the `servername` is used by a multi-homed server to choose the correct certificate to present to the client.

I guessed all kafka hosts were using a same certificate so that I tried to put one of host name in `servername`. Bingo! connected.
 
Hope it helps you.

**Comment by ericqqqqq** [comment]
Thanks @goemondo 
It worked for me. I had to pass the name of the leader broker to `servername`. However, I remember that leader broker changes, maybe not a long-term solution for me.

I plan to use [node-rdkafka](https://github.com/Blizzard/node-rdkafka/) instead. I found that library is officially tested by IBM Event Streams Team. Also found this [document](https://github.com/ibm-cloud-docs/EventStreams/blame/master/eventstreams141.md#L115-L127), not sure why it is commented out.

It is a great workaround, thanks! Helped me a lot.
